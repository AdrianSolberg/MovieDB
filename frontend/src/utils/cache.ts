import { makeVar } from "@apollo/client";
import { Filters, SortingType } from "@/types/movieTypes";

export const sortOptionVar = makeVar<SortingType>(
    (sessionStorage.getItem("sort_option") as SortingType) ??
        SortingType.NEWEST_FIRST
);


const storedFilters = sessionStorage.getItem("filters");
const initialFilters: Filters = storedFilters ? JSON.parse(storedFilters) : {
    Genre: [],
    Rating: [],
    Decade: [],
    Status: [],
    Runtime: [],
};
export const filtersVar = makeVar<Filters>(initialFilters);

export const searchVar = makeVar<string>(
    sessionStorage.getItem("search") ?? ""
);

export const usernameVar = makeVar<string>(
    localStorage.getItem("username") ?? "Guest"
);
