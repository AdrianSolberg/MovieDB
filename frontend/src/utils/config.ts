const BACKEND_PORT = 3001;

const BACKEND_URL =
    import.meta.env.MODE === "production"
        ? `http://it2810-26.idi.ntnu.no:${BACKEND_PORT}`
        : `http://localhost:${BACKEND_PORT}`;

export { BACKEND_URL };
