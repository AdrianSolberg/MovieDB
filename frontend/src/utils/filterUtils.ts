import { Filters, FiltersInput } from "@/types/__generated__/types";

export function getFiltersAsInput(filters: Filters): FiltersInput {
    return {
        Decade: filters.Decade.map((filter) => filter.name),
        Genre: filters.Genre.map((filter) => filter.name),
        Rating: filters.Rating.map((filter) => filter.name),
        Status: filters.Status.map((filter) => filter.name),
        Runtime: filters.Runtime.map((filter) => filter.name),
    };
}
