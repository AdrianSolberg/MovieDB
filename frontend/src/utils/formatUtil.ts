export const formatDate = (date: Date) => {
    return date.toLocaleDateString("en-US", {
        year: "numeric",
        month: "long",
        day: "numeric",
        hour: "2-digit",
        minute: "2-digit",
    });
};

export const formatNumber = (n: number) => {
    return n.toLocaleString("no-NO");
};

export const formatCurrency = (value: number) => {
    return value.toLocaleString("en-US", {
        style: "currency",
        currency: "USD",
        maximumFractionDigits: 0,
    });
};
