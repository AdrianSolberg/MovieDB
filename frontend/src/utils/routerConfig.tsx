import Navbar from "@/components/Navbar";
import ActivityPage from "@/pages/ActivityPage";
import HomePage from "@/pages/HomePage";
import MovieDetailPage from "@/pages/MovieDetailPage";
import MyReviewsPage from "@/pages/MyReviewsPage";
import NotFoundPage from "@/pages/NotFoundPage";
import { Toaster } from "@/shadcn/components/ui/sonner";
import { Outlet, RouteObject } from "react-router-dom";

const routerConfig: RouteObject[] = [
    {
        path: "/",
        element: (
            <>
                <Navbar />
                <div id="main-content" tabIndex={-1}>
                    <Outlet />
                </div>
                <Toaster position="bottom-right" />
            </>
        ),
        errorElement: <NotFoundPage />,
        children: [
            {
                path: "",
                element: <HomePage />,
            },
            {
                path: "movie/:movieId",
                element: <MovieDetailPage />,
            },
            {
                path: "activity",
                element: <ActivityPage />,
            },
            {
                path: "myReviews",
                element: <MyReviewsPage />,
            },
        ],
    },
];

export default routerConfig;
