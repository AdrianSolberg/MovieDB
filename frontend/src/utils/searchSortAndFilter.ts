import { Movie, SortingType } from "@/__generated__/types";

const filterByGenres = (movies: Movie[], selectedGenres: string[]) => {
    if (selectedGenres.length) {
        return movies.filter((movie) =>
            selectedGenres.every((selectedGenre) =>
                movie.genres.some((genre) => genre === selectedGenre)
            )
        );
    }
    return movies;
};

const filterByRating = (movies: Movie[], selectedRatings: string[]) => {
    if (selectedRatings.length) {
        return movies.filter((movie) =>
            selectedRatings.includes(
                Math.round(movie.vote_average / 2).toString()
            )
        );
    }
    return movies;
};

const filterByReleaseYear = (movies: Movie[], selectedDecades: string[]) => {
    if (selectedDecades.length) {
        return movies.filter((movie) => {
            const releaseYear = movie.release_date.getFullYear();
            return selectedDecades.some((decade) => {
                const startYear = parseInt(decade.slice(0, 4));
                return releaseYear >= startYear && releaseYear < startYear + 10;
            });
        });
    }
    return movies;
};

const filterByLanguage = (movies: Movie[], selectedLanguages: string[]) => {
    if (selectedLanguages.length) {
        return movies.filter((movie) =>
            selectedLanguages.every((selectedLanguage) =>
                movie.spoken_languages.some(
                    (language) => language === selectedLanguage
                )
            )
        );
    }
    return movies;
};

const filterByStatus = (movies: Movie[], selectedStatuses: string[]) => {
    if (selectedStatuses.length) {
        return movies.filter((movie) =>
            selectedStatuses.includes(movie.status)
        );
    }
    return movies;
};

const filterByRuntime = (movies: Movie[], selectedRuntimes: string[]) => {
    if (selectedRuntimes.length) {
        return movies.filter((movie) => {
            const runtime = movie.runtime;
            return selectedRuntimes.some((runtimeFilter) => {
                switch (runtimeFilter) {
                    case "Less than 1 hour":
                        return runtime < 60;
                    case "1 - 2 hours":
                        return runtime >= 60 && runtime < 120;
                    case "2 - 3 hours":
                        return runtime >= 120 && runtime < 180;
                    case "3 hours or more":
                        return runtime >= 180;
                    default:
                        return false;
                }
            });
        });
    }
    return movies;
};

export const filterMovies = (
    movies: Movie[],
    filters: { [key: string]: string[] }
) => {
    let filteredMovies = movies;

    const filterFunctions = {
        Genre: filterByGenres,
        Rating: filterByRating,
        "Release Year": filterByReleaseYear,
        Language: filterByLanguage,
        Status: filterByStatus,
        Runtime: filterByRuntime,
    };

    for (const [filterKey, filterFunction] of Object.entries(filterFunctions)) {
        filteredMovies = filterFunction(
            filteredMovies,
            filters[filterKey] || []
        );
    }

    return filteredMovies;
};

const sortingMap: Map<SortingType, (a: Movie, b: Movie) => number> = new Map<
    SortingType,
    (a: Movie, b: Movie) => number
>([
    [
        SortingType.NewestFirst,
        (a, b) => b.release_date.getTime() - a.release_date.getTime(),
    ],
    [
        SortingType.OldestFirst,
        (a, b) => a.release_date.getTime() - b.release_date.getTime(),
    ],
    [SortingType.BestRated, (a, b) => b.vote_average - a.vote_average],
    [SortingType.WorstRated, (a, b) => a.vote_average - b.vote_average],
    [SortingType.LongestRuntime, (a, b) => b.runtime - a.runtime],
    [SortingType.ShortestRuntime, (a, b) => a.runtime - b.runtime],
]);

export const sortMovies = (sortOption: SortingType, movies: Movie[]) => {
    if (sortingMap.has(sortOption)) {
        movies.sort(sortingMap.get(sortOption));
    }

    return movies;
};

export const searchMovies = (search: string, movies: Movie[]) => {
    return movies.filter((movie) =>
        movie.title.toLowerCase().includes(search.toLowerCase())
    );
};

const sortingDisplayNames: Map<SortingType, string> = new Map<
    SortingType,
    string
>([
    [SortingType.BestRated, "Best rated"],
    [SortingType.LongestRuntime, "Longest runtime"],
    [SortingType.NewestFirst, "Newest first"],
    [SortingType.OldestFirst, "Oldest first"],
    [SortingType.ShortestRuntime, "Shortest runtime"],
    [SortingType.WorstRated, "Worst rated"],
]);

export const getSortOptionDisplayName = (sortType: SortingType) => {
    return sortingDisplayNames.get(sortType);
};
