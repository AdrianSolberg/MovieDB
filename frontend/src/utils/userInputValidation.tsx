import { toast } from "sonner";

/**
 *  Returns true if the username is valid, false otherwise.
 *  A valid username must be between 3 and 20 characters long.
 *
 *  Displays a toast error if the username is too short or too long.
 */
export function validateUsername(username: string) {
    const usernameMinLength = 3;
    const usernameMaxLength = 20;

    if (username.length < usernameMinLength) {
        toast.error("Username is too short", {
            description: `Username must be at least ${usernameMinLength} characters long`,
        });
        return false;
    }

    if (username.length > usernameMaxLength) {
        toast.error("Username is too long", {
            description: `Username must be at most  ${usernameMaxLength} characters long`,
        });
        return false;
    }

    return true;
}

/**
 * Returns true if the review is valid, false otherwise.
 * A valid review must be at most 1500 characters long.
 *
 * Displays a toast error if the review is too long.
 */
export function validateReview(comment: string) {
    const reviewMaxLength = 1500;

    if (comment.trim().length > reviewMaxLength) {
        toast.error("Review is too long", {
            description: `Review must be at most ${reviewMaxLength} characters.`,
        });
        return false;
    }

    return true;
}
