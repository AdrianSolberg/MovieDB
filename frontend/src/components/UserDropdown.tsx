import { usernameVar } from "@/utils/cache";
import { Button } from "@/shadcn/components/ui/button";
import {
    Dialog,
    DialogContent,
    DialogDescription,
    DialogHeader,
    DialogTitle,
    DialogTrigger,
} from "@/shadcn/components/ui/dialog";
import {
    DropdownMenu,
    DropdownMenuContent,
    DropdownMenuItem,
    DropdownMenuLabel,
    DropdownMenuSeparator,
    DropdownMenuTrigger,
} from "@/shadcn/components/ui/dropdown-menu";
import { Input } from "@/shadcn/components/ui/input";
import { Label } from "@/shadcn/components/ui/label";
import { validateUsername } from "@/utils/userInputValidation";
import { useReactiveVar } from "@apollo/client";
import { Edit, LogOut, User } from "lucide-react";
import { useState } from "react";
import { toast } from "sonner";

const UserDropdown = () => {
    const username = useReactiveVar(usernameVar);
    const [newUsername, setNewUsername] = useState("");
    const [isDropdownOpen, setIsDropdownOpen] = useState(false);
    const [isDialogOpen, setIsDialogOpen] = useState(false);

    const handleUsernameChange = (e: React.FormEvent) => {
        e.preventDefault();

        const trimmedUsername = newUsername.trim();
        if (!validateUsername(trimmedUsername)) return;

        if (trimmedUsername) {
            usernameVar(trimmedUsername);
            localStorage.setItem("username", trimmedUsername);
            setNewUsername("");
            setIsDialogOpen(false);
            setIsDropdownOpen(false);
            toast.success("Username changed successfully");
        }
    };

    const handleSignOut = () => {
        localStorage.removeItem("username");
        usernameVar("Guest");
        toast.success("Signed out successfully");
    };

    const handleDialogOpenChange = (open: boolean) => {
        setIsDialogOpen(open);
        setNewUsername("");
    };

    return (
        <DropdownMenu open={isDropdownOpen} onOpenChange={setIsDropdownOpen}>
            <DropdownMenuTrigger asChild>
                <Button
                    variant="outline"
                    className="flex items-center space-x-2"
                    aria-label="user"
                >
                    <User className="h-4 w-4" />
                    <span className="hidden truncate sm:flex sm:max-w-32">
                        {username}
                    </span>
                </Button>
            </DropdownMenuTrigger>
            <DropdownMenuContent align="end" className="w-56">
                <DropdownMenuLabel>My Account</DropdownMenuLabel>
                <DropdownMenuSeparator />
                <DropdownMenuLabel className="truncate">
                    {username}
                </DropdownMenuLabel>
                <DropdownMenuSeparator />
                <Dialog
                    open={isDialogOpen}
                    onOpenChange={(open) => handleDialogOpenChange(open)}
                >
                    <DialogTrigger asChild>
                        <DropdownMenuItem onSelect={(e) => e.preventDefault()}>
                            <Edit className="mr-2 h-4 w-4" />
                            Change username
                        </DropdownMenuItem>
                    </DialogTrigger>
                    <DialogContent>
                        <DialogHeader>
                            <DialogTitle>Change username</DialogTitle>
                            <DialogDescription />
                        </DialogHeader>
                        <form
                            onSubmit={handleUsernameChange}
                            className="space-y-4"
                        >
                            <div>
                                <Label htmlFor="new-username">
                                    New username
                                </Label>
                                <Input
                                    id="new-username"
                                    placeholder="Enter new username"
                                    value={newUsername}
                                    onChange={(e) =>
                                        setNewUsername(e.target.value)
                                    }
                                    maxLength={20}
                                />
                            </div>
                            <Button type="submit">Change username</Button>
                        </form>
                    </DialogContent>
                </Dialog>
                <DropdownMenuItem onClick={handleSignOut}>
                    <LogOut className="mr-2 h-4 w-4" />
                    Sign out
                </DropdownMenuItem>
            </DropdownMenuContent>
        </DropdownMenu>
    );
};

export default UserDropdown;
