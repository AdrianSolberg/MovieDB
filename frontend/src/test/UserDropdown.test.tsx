import { render, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import userEvent from "@testing-library/user-event";
import UserDropdown from "@/components/UserDropdown";
import { vi } from "vitest";
import { validateUsername } from "@/utils/userInputValidation";

vi.mock("@/utils/userInputValidation", () => ({
    validateUsername: vi.fn(),
}));

describe("UserDropdown", () => {
    beforeEach(() => {
        vi.mocked(validateUsername).mockReturnValue(true);
    });

    afterEach(() => {
        vi.clearAllMocks();
        localStorage.clear();
    });

    it("matches snapshot", () => {
        const { asFragment } = render(<UserDropdown />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("successfully renders UserDropdown", async () => {
        render(<UserDropdown />);
        expect(screen.getByText("Guest")).toBeInTheDocument();
    });

    it("successfully opens UserDropdown", async () => {
        render(<UserDropdown />);
        await userEvent.click(screen.getByText("Guest"));
        expect(screen.getByText("My Account")).toBeInTheDocument();
        expect(screen.getAllByText("Guest")).toHaveLength(2); // In dropdown button and menu
        expect(screen.getByText("Change username")).toBeInTheDocument();
        expect(screen.getByText("Sign out")).toBeInTheDocument();
    });

    it("successfully closes dropdown on sign out as guest", async () => {
        render(<UserDropdown />);
        await userEvent.click(screen.getByText("Guest"));
        await userEvent.click(screen.getByText("Sign out"));
        expect(screen.queryByText("My Account")).toBeNull();
    });

    it("successfully opens dialog on change username", async () => {
        render(<UserDropdown />);
        await userEvent.click(screen.getByText("Guest"));
        await userEvent.click(screen.getByText("Change username"));
        expect(screen.getByText("New username")).toBeInTheDocument();
    });

    it("successfully changes username and signs out", async () => {
        render(<UserDropdown />);
        await userEvent.click(screen.getByText("Guest"));
        await userEvent.click(screen.getByText("Change username"));

        const input = screen.getByPlaceholderText("Enter new username");
        const newUsername = "TestUser";
        await userEvent.type(input, newUsername);
        await userEvent.click(
            screen.getByRole("button", { name: "Change username" })
        );

        expect(screen.queryByText("New username")).toBeNull(); // Should close dialog
        expect(screen.queryByText("My Account")).toBeNull(); // Should close menu
        expect(screen.getByText(newUsername)).toBeInTheDocument(); // In dropdown button

        await userEvent.click(screen.getByText("TestUser"));
        await userEvent.click(screen.getByText("Sign out"));
        expect(screen.getByText("Guest")).toBeInTheDocument();
        expect(screen.queryByText("My Account")).toBeNull();
    });

    it("successfully handles enter click for username change", async () => {
        render(<UserDropdown />);
        await userEvent.click(screen.getByText("Guest"));
        await userEvent.click(screen.getByText("Change username"));

        const input = screen.getByPlaceholderText("Enter new username");
        const newUsername = "TestUser";
        await userEvent.type(input, `${newUsername}{enter}`);

        expect(screen.queryByText("New username")).toBeNull(); // Should close dialog
        expect(screen.queryByText("My Account")).toBeNull(); // Should close menu
        expect(screen.getByText(newUsername)).toBeInTheDocument(); // In dropdown button

        await userEvent.click(screen.getByText("TestUser"));
        await userEvent.click(screen.getByText("Sign out"));
    });

    it("successfully handles empty string as new username", async () => {
        vi.mocked(validateUsername).mockReturnValue(false);
        render(<UserDropdown />);
        await userEvent.click(screen.getByText("Guest"));
        await userEvent.click(screen.getByText("Change username"));

        const input = screen.getByPlaceholderText("Enter new username");
        const newUsername = "";
        await userEvent.type(input, `${newUsername}{enter}`);

        expect(screen.getByText("New username")).toBeInTheDocument(); // Should not close dialog
    });

    it("successfully clears new username string", async () => {
        render(<UserDropdown />);
        await userEvent.click(screen.getByText("Guest"));
        await userEvent.click(screen.getByText("Change username"));

        const input = screen.getByPlaceholderText("Enter new username");
        const newUsername = "TestUser";
        await userEvent.type(input, newUsername);
        await userEvent.click(screen.getByText("Close"));

        expect(screen.queryByText("New username")).toBeNull(); // Should close dialog
        await userEvent.click(screen.getByText("Change username"));
        expect(screen.queryByText("TestUser")).toBeNull();
    });
});
