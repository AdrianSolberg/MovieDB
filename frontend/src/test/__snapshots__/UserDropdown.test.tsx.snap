// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`UserDropdown > matches snapshot 1`] = `
<DocumentFragment>
  <button
    aria-description="Display your username, change your username or sign out"
    aria-expanded="false"
    aria-haspopup="menu"
    aria-label="user"
    class="justify-center whitespace-nowrap rounded-md text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 border border-input bg-background hover:bg-accent hover:text-accent-foreground h-10 px-4 py-2 flex items-center space-x-2"
    data-state="closed"
    id="radix-:r0:"
    type="button"
  >
    <svg
      class="lucide lucide-user h-4 w-4"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M19 21v-2a4 4 0 0 0-4-4H9a4 4 0 0 0-4 4v2"
      />
      <circle
        cx="12"
        cy="7"
        r="4"
      />
    </svg>
    <span
      class="hidden truncate sm:flex sm:max-w-32"
    >
      Guest
    </span>
  </button>
</DocumentFragment>
`;
